lib/ansible/playbook/task.py

```diff
@@ -144,8 +144,43 @@ def load(data, block=None, role=None, task_include=None, variable_manager=None,
         return task.load_data(data, variable_manager=variable_manager, loader=loader)
 
     def _post_validate_module_defaults(self, attr: str, value: t.Any, templar: TemplateEngine) -> t.Any:
-        """Override module_defaults post validation to disable templating, which is handled by args post validation."""
-        return value
+        """Template and resolve all actions/modules/groups in module_defaults. Argument validation is handled by args post validation."""
+        finalized = []
```

Comment:
This should be at least
```suggestion
        finalized: list[dict[str, object]] = []
```
(as MyPy suggested)

MyPy cannot infer the type because when the initial list is empty, it doesn't know what's inside. And it infers the type when it sees the first initialization, whatever manipulation happen later are checked against what's been inferred or otherwise specified explicitly.